src/Microsoft.AspNetCore.Routing/Internal/UriBuildingContext.cs

```diff
@@ -205,17 +205,24 @@ public override string ToString()
 
         private void EncodeValue(string value)
         {
-            EncodeValue(value, encodeSlashes: true);
+            EncodeValue(value, encodeSlashes: true, parameterTransformer: null);
         }
 
-        private void EncodeValue(string value, bool encodeSlashes)
+        private void EncodeValue(string value, bool encodeSlashes, ParameterTransformer parameterTransformer)
         {
-            EncodeValue(value, start: 0, characterCount: value.Length, encodeSlashes);
+            EncodeValue(value, start: 0, characterCount: value.Length, encodeSlashes, parameterTransformer);
         }
 
         // For testing
-        internal void EncodeValue(string value, int start, int characterCount, bool encodeSlashes)
+        internal void EncodeValue(string value, int start, int characterCount, bool encodeSlashes, ParameterTransformer parameterTransformer)
         {
+            if (parameterTransformer != null)
+            {
+                value = parameterTransformer.Transform(value.Substring(0, characterCount));
+                start = 0;
+                characterCount = value.Length;
+            }
```

Comment:
Hmm this is a little wierd, and maybe I paged out my understanding of this code at a bad time. When is this called with a subsection of a value?

Reply:
It is used here:

```
// Allow the first segment to have a leading slash.
// This prevents the leading slash from PathString segments from being encoded.
if (_path.Length == 0 && value.Length > 0 && value[0] == '/')
{
    _path.Append("/");
    EncodeValue(value, 1, value.Length - 1, encodeSlashes, parameterTransformer);
}
else
{
    EncodeValue(value, encodeSlashes, parameterTransformer);
}
```